coveo-functools/README.md

```diff
@@ -102,15 +103,86 @@ Flex can be used with:
 ### Limitations
 
 - Variable positional args (such as `def fn(*args): ...`) are left untouched.
-- Basic json-compatible types are left untouched. This is determined by the annotation, not the actual value.
+- Basic json-compatible types will be left untouched. This is determined by the annotation, not the actual value.
 - If `None` is given as a value to deserialize into anything, `None` is given back. Absolutely no validation occurs in this case.
-
-- You can only `Union` basic json-compatible types, or `List[T], T`
+- An Abstract class requires a subclass adapter that returns the non-abstract class to use.
 - No support for additional `typing` and `collections` objects other than the ones mentioned in this documentation.
+- You can only `Union` basic json-compatible types, or `List[T], T`.
+
 
 These are subject to change.
 
 
+### Subclass adapters
+
+To use Abstract classes as annotations, you need to register subclass adapters.
+
+The adapter is a `Callable[[Any], TypeHint]` that you provide. 
+It will be called with the payload value as `Any`, so you can inspect the content.
+It must return a `TypeHint` that tells flex which class to use.
+
+With subclass adapters, you can selectively decide the implementation class based on the payload to deserialize.
+While this is necessary when annotating structures with Abstract classes, 
+it can be used for any other class as well.
+
+For this to work, you must register the annotated class with a callback:
+
+```python
+from coveo_functools.flex.subclass_adapter import register_subclass_adapter
+
+class Abstract:
+  @abstractmethod
+  def api(self) -> None:
+    ...
+
+  
+class ThisImplementation(Abstract):
+  def api(self) -> None:
+    ...
+  
+  
+class OtherImplementation(Abstract):
+  def api(self) -> None:
+    ...
+  
+  
+def adapter(payload: Any) -> Type:
+  assert isinstance(payload, dict)  # actual type depends on payload
+  return ThisImplementation if 'this' in payload else OtherImplementation
+  
+  
+register_subclass_adapter(Abstract, adapter)
+```
+
+Thanks to the adapter, this is now possible:
+
+```python
+assert isinstance(deserialize({'this': {}}, hint=Abstract), ThisImplementation)
+assert isinstance(deserialize({}, hint=Abstract), OtherImplementation)
+
+
+@dataclass
+class Payload:
+    owner: Abstract
+    
+
+instance = deserialize({"owner": {"this": {}}}, hint=Payload)
+assert isinstance(instance, ThisImplementation)
+```
+
+The intended use of subclass adapters is to support Abstract classes as annotations.
+
+Any other use will generally mess up your type annotation game because they alter the types dynamically at runtime.
+This will never play well with IDEs and linters which rely on static analysis.
+
+
+Keep in mind that:
+  - There are no validations, to allow duck typing and stuff

```

Comment:
weird ponctuation ponctuation

Reply:
explain yourself.

Reply:
I missed it myself but the comma should not be there. It's not an option btw. ;-)

Reply:
I read your sentence 10 more times and I think I know what you mean and you are actually right. :P The better phrasing would be to change the comma for 'in order to allow :duck: typing and stuff`

Reply:
Extra point if you the duck emoji in the text

Reply:
```suggestion
  - There are no validations (to allow duck typing and stuff)

```